src/model/neuroevolution/neuroevolution.py

```diff
@@ -128,17 +129,19 @@ def _mutation(self, individuals: List[ChildIndividual]) -> List[ChildIndividual]
                 )(individual)
         return individuals
 
-    def evolve(self) -> None:
+    def evolve(self, with_parents: bool) -> None:
         adaptations = self._environment.compute_adaptations(
-            nn.neural_network for nn in self._new_generation
+            (child.neural_network for child in self._new_generation),
+            (parent.neural_network for parent in self.individuals)
+            if with_parents
+            else [],
         )
         new_individuals = [
-            Individual(i.neural_network, adaptation)
-            for i, adaptation in zip(self._new_generation, adaptations)
+            Individual(child.neural_network, adaptation)
+            for child, adaptation in zip(self._new_generation, adaptations)
         ]
         self.individuals.extend(new_individuals)
-        self._sort_individuals()
-        parents: List[Individual] = self._selection()
-        self.individuals = parents
-        children: List[ChildIndividual] = self._reproduction(parents)
+        self._sort_individuals_and_kill_unnecessary()
+        self._parents = self._selection()
+        children: List[ChildIndividual] = self._reproduction(self._parents)
```

Comment:
There is no reason why parents should be a class field. Please also remove self._parents from init.
```suggestion
        parents: List[Individual] = self._selection()
        children: List[ChildIndividual] = self._reproduction(parents)
```

Reply:
Class field _parents is necessary, I forgot to use it before

Reply:
Why is it necessary?

Reply:
In current implementation it's impossible to 'evaluate' whole population.

Reply:
Ok, now I see